driver-core/src/main/com/mongodb/connection/netty/NettyStream.java

```diff
@@ -445,47 +519,52 @@ public void operationComplete(final ChannelFuture future) {
+            return Math.addExact(timeout, additionalTimeout);
+        }
     }
 
-    private void adjustTimeout(final boolean disable, final int additionalTimeout) {
-            if (isClosed) {
-                return;
-            }
-            ChannelHandler timeoutHandler = channel.pipeline().get(READ_HANDLER_NAME);
-            if (timeoutHandler != null) {
-                final ReadTimeoutHandler readTimeoutHandler = (ReadTimeoutHandler) timeoutHandler;
-                final ChannelHandlerContext handlerContext = channel.pipeline().context(timeoutHandler);
-                EventExecutor executor = handlerContext.executor();
-
-                if (disable) {
-                    if (executor.inEventLoop()) {
-                        readTimeoutHandler.removeTimeout(handlerContext);
-                    } else {
-                        executor.submit(new Runnable() {
-                            @Override
-                            public void run() {
-                                readTimeoutHandler.removeTimeout(handlerContext);
-                            }
-                        });
-                    }
-                } else {
-                    if (executor.inEventLoop()) {
-                        readTimeoutHandler.scheduleTimeout(handlerContext, additionalTimeout);
-                    } else {
-                        executor.submit(new Runnable() {
-                            @Override
-                            public void run() {
-                                readTimeoutHandler.scheduleTimeout(handlerContext, additionalTimeout);
-                            }
-                        });
-                    }
+    private static ScheduledFuture<?> scheduleReadTimeout(@Nullable final ReadTimeoutTask readTimeoutTask, final long timeoutMillis) {
+        if (timeoutMillis == NO_SCHEDULE_TIME) {
+            return null;
+        } else {
+            //assert readTimeoutTask != null : "readTimeoutTask must be initialized if read timeouts are enabled";
+            return readTimeoutTask.schedule(timeoutMillis);
+        }
+    }
+
+    @ThreadSafe
+    private static final class ReadTimeoutTask implements Runnable {
+        private final ChannelHandlerContext ctx;
+
+        private ReadTimeoutTask(final ChannelHandlerContext timeoutChannelHandlerContext) {
+            ctx = timeoutChannelHandlerContext;
+        }
+
+        @Override
+        public void run() {
+            try {
+                if (ctx.channel().isOpen()) {
+                    ctx.fireExceptionCaught(ReadTimeoutException.INSTANCE);
+                    ctx.close();
                 }
+            } catch (final Throwable t) {
+                ctx.fireExceptionCaught(t);
             }
+        }
+
+        private ScheduledFuture<?> schedule(final long timeoutMillis) {
+            //assert timeoutMillis > 0 : timeoutMillis;
+            return ctx.executor().schedule(this, timeoutMillis, MILLISECONDS);
+        }
     }
-}
+}
```

Comment:
Newline required at end, I believe

Reply:
Our `.editorconfig` contains `insert_final_newline = false`, so Idea does not allow me to end a file with a newline character. Should I also change the `.editorconfig`?

Reply:
That's strange, since most (all?) existing code ends in a newline.  Since this file did end in a newline let's add it back, and we can figure out the .editorconfig issue separately